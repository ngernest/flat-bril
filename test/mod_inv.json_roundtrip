{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "p",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v0"
          ],
          "dest": "two",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "p"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "two"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v3"
          ],
          "dest": "m",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v4",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v4"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v5"
          ],
          "dest": "a",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v7",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v7"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.6"
        },
        {
          "args": [
            "m"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v9",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v8",
            "v9"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "v10"
          ],
          "funcs": [],
          "labels": [
            "for.body.6",
            "for.end.6"
          ],
          "op": "br"
        },
        {
          "label": "for.body.6"
        },
        {
          "args": [
            "m"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "m"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "two"
          ],
          "dest": "v14",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v13",
            "v14"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "two"
          ],
          "dest": "v16",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v15",
            "v16"
          ],
          "dest": "v17",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v12",
            "v17"
          ],
          "dest": "v18",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v18"
          ],
          "funcs": [],
          "labels": [
            "then.11",
            "else.11"
          ],
          "op": "br"
        },
        {
          "label": "then.11"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "endif.11"
          ],
          "op": "jmp"
        },
        {
          "label": "else.11"
        },
        {
          "args": [
            "ans"
          ],
          "dest": "v19",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "a"
          ],
          "dest": "v20",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v19",
            "v20"
          ],
          "dest": "v21",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "p"
          ],
          "dest": "v22",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v21",
            "v22"
          ],
          "dest": "v23",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v23"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "endif.11"
        },
        {
          "args": [
            "a"
          ],
          "dest": "v24",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "a"
          ],
          "dest": "v25",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v24",
            "v25"
          ],
          "dest": "v26",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "p"
          ],
          "dest": "v27",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v26",
            "v27"
          ],
          "dest": "v28",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "v28"
          ],
          "dest": "a",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "m"
          ],
          "dest": "v29",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "two"
          ],
          "dest": "v30",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v29",
            "v30"
          ],
          "dest": "v31",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v31"
          ],
          "dest": "m",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.6"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.6"
        },
        {
          "args": [
            "ans"
          ],
          "dest": "v32",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v32"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v33",
          "op": "const",
          "type": "int",
          "value": 0
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "p",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "p"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1",
            "v2"
          ],
          "dest": "v3",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "p"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v3",
            "v4"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v0",
            "v5"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v6"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    }
  ]
}
