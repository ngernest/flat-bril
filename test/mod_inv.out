@main
	op: const		dest: v0		type: int		value: 2	
	op: id   		dest: two		type: int		args: ["v0"]	
	op: id   		dest: v1		type: int		args: ["p"]	
	op: id   		dest: v2		type: int		args: ["two"]	
	op: sub  		dest: v3		type: int		args: ["v1", "v2"]	
	op: id   		dest: m 		type: int		args: ["v3"]	
	op: const		dest: v4		type: int		value: 1	
	op: id   		dest: ans		type: int		args: ["v4"]	
	op: id   		dest: v5		type: int		args: ["n"]	
	op: id   		dest: a 		type: int		args: ["v5"]	
	op: const		dest: v7		type: int		value: 1	
	op: id   		dest: i 		type: int		args: ["v7"]	
.for.cond.6
	op: id   		dest: v8		type: int		args: ["m"]	
	op: const		dest: v9		type: int		value: 0	
	op: gt   		dest: v10		type: bool		args: ["v8", "v9"]	
	op: br   		args: ["v10"]		labels: ["for.body.6", "for.end.6"]
.for.body.6
	op: id   		dest: v12		type: int		args: ["m"]	
	op: id   		dest: v13		type: int		args: ["m"]	
	op: id   		dest: v14		type: int		args: ["two"]	
	op: div  		dest: v15		type: int		args: ["v13", "v14"]	
	op: id   		dest: v16		type: int		args: ["two"]	
	op: mul  		dest: v17		type: int		args: ["v15", "v16"]	
	op: eq   		dest: v18		type: bool		args: ["v12", "v17"]	
	op: br   		args: ["v18"]		labels: ["then.11", "else.11"]
.then.11
	op: jmp  		labels: ["endif.11"]
.else.11
	op: id   		dest: v19		type: int		args: ["ans"]	
	op: id   		dest: v20		type: int		args: ["a"]	
	op: mul  		dest: v21		type: int		args: ["v19", "v20"]	
	op: id   		dest: v22		type: int		args: ["p"]	
	op: call 		dest: v23		type: int		args: ["v21", "v22"]		funcs: ["mod"]
	op: id   		dest: ans		type: int		args: ["v23"]	
.endif.11
	op: id   		dest: v24		type: int		args: ["a"]	
	op: id   		dest: v25		type: int		args: ["a"]	
	op: mul  		dest: v26		type: int		args: ["v24", "v25"]	
	op: id   		dest: v27		type: int		args: ["p"]	
	op: call 		dest: v28		type: int		args: ["v26", "v27"]		funcs: ["mod"]
	op: id   		dest: a 		type: int		args: ["v28"]	
	op: id   		dest: v29		type: int		args: ["m"]	
	op: id   		dest: v30		type: int		args: ["two"]	
	op: div  		dest: v31		type: int		args: ["v29", "v30"]	
	op: id   		dest: m 		type: int		args: ["v31"]	
	op: jmp  		labels: ["for.cond.6"]
.for.end.6
	op: id   		dest: v32		type: int		args: ["ans"]	
	op: print		args: ["v32"]	
	op: const		dest: v33		type: int		value: 0	
@mod
	op: id   		dest: v0		type: int		args: ["n"]	
	op: id   		dest: v1		type: int		args: ["n"]	
	op: id   		dest: v2		type: int		args: ["p"]	
	op: div  		dest: v3		type: int		args: ["v1", "v2"]	
	op: id   		dest: v4		type: int		args: ["p"]	
	op: mul  		dest: v5		type: int		args: ["v3", "v4"]	
	op: sub  		dest: v6		type: int		args: ["v0", "v5"]	
	op: ret  		args: ["v6"]	
